File: methods/filter.py
import time


def filter_by_author(quotes, author):
    """
    Filters quotes by author
    :param quotes: The list of quotes to be filtered
    :param author: The inputted author to filter the quotes by
    :return:
        - quotes: A list of quotes by the inputted author
    """
    return [quote for quote in quotes if quote[1].lower() == author.lower()] if author else quotes


def filter_to_daily_quote(quotes):
    """
    Filters quotes to get a quote based on the current day
    :param quotes: The list of quotes to be filtered
    :return:
        - quotes: The filtered list of quotes with a single quote inside
    """
    current_day = int(time.time()) / 86400
    return [quotes[int(current_day) % len(quotes)]]

def filter_by_excluded_ids(quotes, exclude_indexes):
    """
    Filters quotes by excluded indexes
    :param quotes: The list of quotes to be filtered
    :param exclude_indexes: The original string of numbers to be excluded
    :return:
        - quotes: The filtered list of quotes with a single quote inside
    """
    exclude_indexes = [int(index) for index in exclude_indexes.split(",") if index.strip()]
    return [quote for quote in quotes if quote[2] not in exclude_indexes]


def get_specific_quote(quotes, specific_quote):
    """
    Filters quotes to get a specific quote
    :param quotes: The list of quotes to be filtered
    :param specific_quote: The specific quote to be filtered
    :return:
        - quotes: The filtered list of quotes with a single quote inside
    """
    return [quote for quote in quotes if quote[0][:len(specific_quote)].lower() == specific_quote.lower()]


def get_quote_by_index(quotes, specific_quote_index):
    """
    Filters quotes to get a specific quote by index
    :param quotes: The list of quotes to be filtered
    :param specific_quote_index: The index of the specific quote to be filtered
    :return:
        - quotes: The filtered list of quotes with a single quote inside
    """
    return [quote for quote in quotes if quote[2] == specific_quote_index]

File: methods/filter.mjs
import sqlite3 from 'sqlite3';
import { open } from 'sqlite';
import path from 'path';
import { fileURLToPath } from 'url';
import { dirname } from 'path';
import shuffle from './shuffle.mjs';

const __dirname = dirname(fileURLToPath(import.meta.url));

/**
 * Returns an initial unfiltered list of quote objects, each with an id, quote, and author, used for initial error
 * handling
 * @returns {Promise<Array<{id: number, quote: string, author: string}>>} a list of quote objects
 */
export async function allQuotes() {
    const db = await open({
        filename: path.join(__dirname, '../data/quotes.db'),
        driver: sqlite3.Database
    });
    return await db.all('SELECT * FROM quotes');
}

/**
 * Returns a single quote object from the list of quotes
 * @param quoteArray array with quote objects
 * @param seed  seed array
 * @returns {[*]}
 */
const dailyQuoteFilter = (quoteArray, seed) => {
    if(seed !== undefined){
        quoteArray = shuffle(quoteArray, seed);
    }
    const now = new Date();
    const start = new Date('2005-01-20');
    const differenceInTime = now.getTime() - start.getTime();
    const differenceInDays = Math.floor(differenceInTime / (1000 * 3600 * 24));
    return [quoteArray[differenceInDays % quoteArray.length]];
}


/**
 *
 * @param quoteArray the array of all quote objects
 * @param author the name of the author to filter by
 * @param dailyQuote boolean value to determine if the quote is a daily quote
 * @param searchQuote the specific quote we want to search for
 * @param includeIDs the indexes we want to include in generation
 * @param excludeIDs the indexes we want to exclude from generation
 * @returns {Promise<Array<{id: number, quote: string, author: string}>>}
 */
export async function filter(quoteArray, author, dailyQuote, searchQuote, includeIDs, excludeIDs, dailySeed){
    let quotes = quoteArray;
    if(dailyQuote){
        return dailyQuoteFilter(quotes, dailySeed);
    }
    if(author){
        quotes = quotes.filter(quote => quote.author.toLowerCase() === author.toLowerCase());
    }
    if(searchQuote){
        quotes = quotes.filter(quote => quote.quote.substring(0, searchQuote.length).toLowerCase() === searchQuote.toLowerCase());
    }
    if(includeIDs){
        let indexes = includeIDs.split(",").map(index => parseInt(index.trim()));
        quotes = quotes.filter(quote => indexes.includes(quote.id));
    }
    if(excludeIDs){
        let indexes = excludeIDs.split(",").map(index => parseInt(index.trim()));
        quotes = quotes.filter(quote => !indexes.includes(quote.id));
    }
    return quotes;
}